sdks/python/apache_beam/runners/worker/opcounters.py

```diff
@@ -24,11 +24,119 @@
 import math
 import random
 
+from apache_beam.utils import counters
 from apache_beam.utils.counters import Counter
+from apache_beam.utils.counters import CounterName
 
 # This module is experimental. No backwards-compatibility guarantees.
 
 
+class TransformIOCounter(object):
+  """Class to track time and bytes consumed while reading from IO.
+
+  Subclasses should be able to track consumption of IO across steps
+  in the same stage - for instance, if a Shuffle or Side Input iterable
+  is passed down to a next step.
+
+  Some examples of IO can be side inputs, shuffle, or streaming state.
+  """
+
+  def update_current_step(self):
+    """Update the current step within a stage as it may have changed.
+
+    If the state changed, it would mean that an initial step passed a
+    data-accessor (such as a side input / shuffle Iterable) down to the
+    next step in a stage.
+    """
+    pass
```

Comment:
<!--thread_id:cc_163696527_t; commit:b68f9f26428de8ed2f34c0919ee91a1662402d86; resolved:0-->
<!--section:context-quote-->
> **aaltay** wrote:
> Should we instead raise a not implemented error for the base methods?

<!--section:body-->
I plan to use the interface as a No-op counter further down the line. I can raise the exceptions here, and make a No-op subclass if you'd rather have it that way.